### PR TITLE
Added support for cause

### PR DESCRIPTION
**Pull request by DanieleFedeli**
#### Checklist

This PR target #112. It will be possible to create an error with cause.

Typescript and Documentation changes will be applied after the main idea is approved.

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Uzlopak** [comment]
I dont think that this is an appropriate implementation. Cause is usually an optional parameter.

**Comment by erfanium** [comment]
`new FastifyError(new Error('casue'))` syntax doesn't looks good to me. I think it's better to follow actual JavaScript implmentation:
```ts
new FastifyError('message', {
  casue: new Error('casue')
})
```

i'm also okay with this idea:
```ts
new FastifyError({
  message: 'message',
  casue: new Error('casue')
})
```

**Comment by erfanium** [comment]
Also `hasCause` option for `createError` method is not really needed.
I believe it can be done dynamic

**Comment by erfanium** [comment]
@DanieleFedeli I have no idea, but i think it's okay to release a major version.

**Comment by climba03003** [comment]
> I am thinking how to make this backward compatible since the function `FastifyError` accepts a rest arg, maybe it will be needed to change the interface of the function accepting an object instead of positional parameters. Any idea?

It doesn't really matter for the `...rest`.
```js
function Error(...args) {
  if(args.length && typeof args.at(-1) === 'object' && 'cause' in args.at(-1)) {
    this.cause = args.pop()
  }
}
```

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by DanieleFedeli** [comment]
Should I fix for node 14?

**Comment by Uzlopak** [comment]
Yes, please.

File: index.js [hunk 1]
```diff
@@ -17,10 +17,15 @@ function createError (code, message, statusCode = 500, Base = Error) {
     if (!new.target) {
       return new FastifyError(...args)
     }
+
     this.code = code
     this.name = 'FastifyError'
     this.statusCode = statusCode
 
+    if (args.length && typeof args.at(-1) === 'object' && 'cause' in args.at(-1)) {
```
Comment (Uzlopak) [inline index.js:25]:
```suggestion
    const lastElement = args.length - 1
    if (lastElement !== -1 && args[lastElement] && typeof args[lastElement] === 'object' && 'cause' in args[lastElement]) {
```

I added the args[lastElement] for the case that the last element is null.

Comment (Uzlopak) [reply index.js:25]: not tested.

**Review by gurgunday — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
@Uzlopak could you do a final check?

**Review by Uzlopak — APPROVED** [review]
LGTM

**Comment by Uzlopak** [comment]
@mcollina :)

**Pull request merged**